code/modules/overmap/contacts/contact_sensors.dm

```diff
@@ -0,0 +1,150 @@
+#define SENSORS_DISTANCE_COEFFICIENT 5
+/obj/machinery/computer/ship/sensors
+	var/list/objects_in_view = list() // Associative list of objects in view -> identification process
+	var/list/contact_datums = list()
+	var/list/trackers = list()
+
+/obj/machinery/computer/ship/sensors/Destroy()
+	objects_in_view.Cut()
+	trackers.Cut()
+
+	for(var/key in contact_datums)
+		var/datum/overmap_contact/record = contact_datums[key]
+		qdel(record)
+	contact_datums.Cut()
+	. = ..()
+
+/obj/machinery/computer/ship/sensors/attempt_hook_up(obj/effect/overmap/visitable/ship/sector)
+	. = ..()
+	if(. && linked && !contact_datums[linked])
+		var/datum/overmap_contact/record = new(src, linked)
+		contact_datums[linked] = record
+		record.marker.alpha = 255
+
+/obj/machinery/computer/ship/sensors/proc/reveal_contacts(var/mob/user)
+	if(user && user.client)
+		for(var/key in contact_datums)
+			var/datum/overmap_contact/record = contact_datums[key]
+			if(record)
+				user.client.images |= record.marker
+
+/obj/machinery/computer/ship/sensors/proc/hide_contacts(var/mob/user)
+	if(user && user.client)
+		for(var/key in contact_datums)
+			var/datum/overmap_contact/record = contact_datums[key]
+			if(record)
+				user.client.images -= record.marker
+
+/obj/machinery/computer/ship/sensors/Process()
+	..()
+	update_sound()
+	if(!linked)
+		return
+	
+	// Update our own marker icon regardless of power or sensor connections.
+	var/sensor_range = round(sensors.range,1)
+	var/datum/overmap_contact/self_record = contact_datums[linked]
+	self_record.update_marker_icon(sensor_range)
+	self_record.show()
+
+	// Update our 'sensor range' (ie. overmap lighting)
+	if(!sensors || !sensors.use_power || !sensors.powered())
+		for(var/key in contact_datums)
+			var/datum/overmap_contact/record = contact_datums[key]
+			if(record.effect == linked)
+				continue
+			qdel(record) // Immediately cut records if power is lost.
+
+		objects_in_view.Cut()
+		return
+
+	// What can we see?
+	var/list/objects_in_current_view = list()
+
+	// Find all sectors with a tracker on their z-level. Only works on ships when they are in space.
+	for(var/obj/item/ship_tracker/tracker in trackers)
+		if(tracker.enabled)
+			var/obj/effect/overmap/visitable/tracked_effect = map_sectors["[get_z(tracker)]"]
+			if(tracked_effect && istype(tracked_effect) && tracked_effect != linked && tracked_effect.requires_contact)
+				objects_in_current_view[tracked_effect] = TRUE
+				objects_in_view[tracked_effect] = 100
+
+	for(var/obj/effect/overmap/contact in (view(sensor_range, linked)-linked))
```

Comment:
There are some kind of dm-level optimizations for iterating view directly without doing this list op which in some cases are very dramatic (unsure about this one). I'd move it to the logic below instead of here.